### PR TITLE
[BUG] Handle capacity usage per epoch for reconnection service

### LINKED ISSUE #74: [bug] Capacity calculations do not track actual capacity usage by service
Problem: The reconnection service allows for specification of a capacity threshold (as either a percentage or a total amount) of available capacity that it is not allowed to exceed (in order that the service should not use up all of a provider's capacity). Currently the code uses `remainingCapacity` vs. `totalCapacityIssued` to make this determination; however, this does not account for other applications' use of capacity.

We need to track the amount of capacity actually used by the reconnection service in the current epoch in order to correctly determine whether the capacity threshold has been exceeded.

### PR DESCRIPTION
**Pull request by saraswatpuneet**
Closes #74

### CODE REVIEW

File: src/blockchain/blockchain.service.ts [hunk 1]
```diff
@@ -107,15 +107,17 @@ export class BlockchainService implements OnApplicationBootstrap, OnApplicationS
...
     const providerU64 = this.api.createType('u64', providerId);
     const { epochStart }: PalletCapacityEpochInfo = await this.query('capacity', 'currentEpochInfo');
     const epochBlockLength: u32 = await this.query('capacity', 'epochLength');
     const capacityDetailsOption: Option<PalletCapacityCapacityDetails> = await this.query('capacity', 'capacityLedger', providerU64);
     const { remainingCapacity, totalCapacityIssued } = capacityDetailsOption.unwrapOr({ remainingCapacity: 0, totalCapacityIssued: 0 });
     const currentBlock: u32 = await this.query('system', 'number');
-
+    const currentEpoch: u32  = await this.query('capacity', 'currentEpoch');
     return {
+      currentEpoch: typeof currentEpoch === 'number' ? BigInt(currentEpoch) : currentEpoch.toBigInt(),
```
Comment (saraswatpuneet) [inline src/blockchain/blockchain.service.ts:120]: also return current epoch

File: src/processor/queue-consumer.service.ts [hunk 2]
```diff
@@ -164,38 +168,46 @@ export class QueueConsumerService extends WorkerHost implements OnApplicationBoo
...
+
   private async checkCapacity(): Promise<void> {
     const capacityLimit = this.configService.getCapacityLimit();
     const capacity = await this.blockchainService.capacityInfo(this.configService.getProviderId());
     const remainingCapacity = capacity.remainingCapacity;
-    let outOfCapacity = capacity.remainingCapacity <= 0n;
+    const currentEpoch = capacity.currentEpoch;
+    const epochCapacityKey = `epochCapacity:${currentEpoch}`;
+    const epochUsedCapacity = BigInt(await this.cacheManager.get(epochCapacityKey) ?? 0); // Fetch capacity used by the service
+    let outOfCapacity = remainingCapacity <= 0n || epochUsedCapacity >= remainingCapacity;
```
Comment (saraswatpuneet) [inline src/processor/queue-consumer.service.ts:187]: if there is no remaining capacity or epochCapacity passed the remaining capacity barrier due the other outside services consuming the capacity too so we pause

Comment (saraswatpuneet) [reply src/processor/queue-consumer.service.ts:187]: this is where use of remainingCapacity makes sense

File: src/processor/queue-consumer.service.ts [hunk 3]
```diff
@@ -164,38 +168,46 @@ export class QueueConsumerService extends WorkerHost implements OnApplicationBoo
...
-              }))`,
-            );
-          }
+            const capacityLimitPercentage = BigInt(capacityLimit.value);
+            const capacityLimitThreshold = (capacity.totalCapacityIssued * capacityLimitPercentage) / 100n;
+            this.logger.debug(`Capacity limit threshold: ${capacityLimitThreshold}`);
+            if (epochUsedCapacity >= capacityLimitThreshold) {
+                outOfCapacity = true;
+                this.logger.warn(`Capacity threshold reached: used ${epochUsedCapacity} of ${capacityLimitThreshold}`);
+            }
```
Comment (saraswatpuneet) [inline src/processor/queue-consumer.service.ts:198]: if `epochCapacity` exceeds the % of total allowed then set `outOfCapacity` to be true

File: src/processor/queue-consumer.service.ts [hunk 4]
```diff
@@ -164,38 +168,46 @@ export class QueueConsumerService extends WorkerHost implements OnApplicationBoo
...
   }
 
+  private async setEpochCapacity(totalCapacityUsed: bigint): Promise<void> {
+    const capacityInfo = await this.blockchainService.capacityInfo(this.configService.getProviderId());
+    const currentEpoch = capacityInfo.currentEpoch;
+    const epochCapacityKey = `epochCapacity:${currentEpoch}`;
+    const epochCapacity = BigInt(await this.cacheManager.get(epochCapacityKey) ?? 0);
+    const newEpochCapacity = epochCapacity + totalCapacityUsed;
+    await this.cacheManager.set(epochCapacityKey, newEpochCapacity.toString());
+  }
```
[the remark this answers is not in the document]
Comment (JoeCap08055) [reply src/processor/queue-consumer.service.ts:178]: We can set the key to expire after a couple of days, that should be fine.

**Comment by JoeCap08055** [comment]
looks good, but one potential edge case: is it possible that either
* multiple transactions in a single batch are not all in the same block, and cross an epoch boundary? (not sure if this is possible)
* multiple batches cross an epoch boundary <-- more likely than ☝🏻

**Comment by saraswatpuneet** [comment]
@JoeCap08055 
I think the second one is more likely to happen, first one i don't think its possible to make that happen considering we make a single extrinsic call per batch and all sub calls are handled by respective pallet 

So for second one, I can retrieve the `epochNumber` before submitting the batch and return that `epochNumber` along side used capacity, that way we know close to possible reality which epoch that transaction went into? You think that would work?

File: src/processor/queue-consumer.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply src/processor/queue-consumer.service.ts:179]: setting the ex to current epoch duration gotten from chain state

**Comment by saraswatpuneet** [comment]
I put this logic to have some approximate number around which a batch was submitted , that way we sample out epochs separately and have separate counter for each, reducing overlap enough to possibly handle the edge case?


https://github.com/AmplicaLabs/reconnection-service/pull/75/commits/e8fe26f9eb769257560f118df7cf4d92d6022a11

File: src/processor/queue-consumer.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoeCap08055) [reply src/processor/queue-consumer.service.ts:179]: Hmm, isn't epoch duration/length just number of blocks? Probably need to convert that to seconds. I also think we should increase the expiration by a factor, just to guarantee it won't expire before we're done with it. It's okay if it hangs around a while past when we need it; we just don't want the cache to get overly polluted with old keys.

[the remark this answers is not in the document]
Comment (saraswatpuneet) [reply src/processor/queue-consumer.service.ts:180]: Good catch, left over from testing

**Comment by saraswatpuneet** [comment]
merging it in other PR for testing and finalization

**Comment by saraswatpuneet** [comment]
Merging this one  into other PR as it is derived from there.

**Pull request merged**